Fix coor api stride

marscher opened the pull request:
* [tica] fix handling of stride in estimate

   Also do not store running covar as instance, because it is only needed
   during estimation.
* [coor/api] 
   * removed _param_stage and _get_input_stage (handled by estimate)
   * all chunksizes default to None (Impl chooses a value, eg. Iterable has a global setting currently 5000). This unifies the mess of different default values in the api.
* [streaming_estimator] chunksize is parameter of estimate, which passes it to the input iterable.

marscher committed:
[tica] fix handling of stride in estimate

Also do not store running covar as instance, because it is only needed
during estimation.

marscher committed:
[coor/api] removed stride param from _param_stage (handled by estimate)

marscher committed:
[streaming_estimator] propagate chunksize to input iterable

marscher committed:
fix test

marscher committed:
dont need to run to test chunksize

marscher committed:
[tica] fix partial_fit (set covar to none)

marscher committed:
[Iterable] if passed chunksize is None, set to 1000

marscher committed:
[lagged covar] chunksize is not an estimation parameter.

marscher committed:
[coordinates] chunksize defaults to None, added DEFAULT_CHUNKSIZE param.

Removed implementation detail from api (wiring data producers, set cs)
from api to StreamingTransformer.estimate().

fixed docstrings in API.

marscher committed:
removed ctor param cs from streaming estimator (estimtate param)

marscher committed:
cleanup

clonker commented on pyemma/coordinates/data/_base/iterable.py:
```diff
@@ -20,16 +20,18 @@
 from pyemma.util.types import is_int
 
+# this is used, in case None is passed as input chunk size.
+DEFAULT_CHUNKSIZE = 5000
```
as we've discussed, might be beneficial to configure this value via the config file

clonker commented on pyemma/coordinates/data/_base/iterable.py:
```diff
@@ -20,16 +20,18 @@
 class Iterable(Loggable, metaclass=ABCMeta):
 
-    def __init__(self, chunksize=1000):
+    def __init__(self, chunksize=DEFAULT_CHUNKSIZE):
```
i believe this is inconsistent, the default chunk size is set if the argument is `None`, however the argument itself defaults to the default chunksize

clonker replied:
also it is called `chunksize` instead of `chunk_size`

clonker commented on pyemma/coordinates/data/_base/transformer.py:
```diff
@@ -115,7 +115,7 @@ class StreamingTransformer(Transformer, DataSource, NotifyOnChangesMixIn):
 
     """
-    def __init__(self, chunksize=1000):
+    def __init__(self, chunksize=None):
```
similar to the previous remark: here you default the argument to None (which probably is what is desired)

clonker commented on pyemma/coordinates/api.py:
```diff
@@ -364,7 +366,7 @@ def source(inp, features=None, top=None, chunk_size=None, **kw):
 
 
-def combine_sources(sources, chunksize=1000):
+def combine_sources(sources, chunksize=None):
```
here it's called `chunksize` instead of `chunk_size`

clonker commented on pyemma/coordinates/api.py:
```diff
@@ -628,7 +639,7 @@ def save_traj(traj_inp, indexes, outfile, top=None, stride = 1, chunksize=1000,
         in :py:obj:`traj_inp.trajfiles`.
 
-    chunksize : int. Default 1000.
+    chunksize : int. Default=1000.
```
here it's called `chunksize` instead of `chunk_size`

clonker commented on pyemma/coordinates/api.py:
```diff
@@ -1012,12 +991,14 @@ def pca(data=None, dim=-1, var_cutoff=0.95, stride=1, mean=None, skip=0):
-    return _param_stage(data, res, stride=stride)
+    res = PCA(dim=dim, var_cutoff=var_cutoff, mean=None, skip=skip, stride=stride)
+    if data is not None:
+        res.estimate(data, chunksize=chunk_size)
```
`chunksize` or `chunk_size`?

clonker commented on pyemma/coordinates/api.py:
```diff
@@ -1215,7 +1202,7 @@ def tica(data=None, lag=10, dim=-1, var_cutoff=0.95, kinetic_map=True, commute_m
                 raise ValueError("Koopman re-weighting is designed for reversible processes, set reversible=True")
             koop = _KoopmanEstimator(lag=lag, stride=stride, skip=skip, ncov_max=ncov_max)
-            _param_stage(data, koop, stride=stride)
+            koop.estimate(data, chunksize=chunk_size)
```
`chunk_size`

clonker commented on pyemma/coordinates/api.py:
```diff
@@ -1239,20 +1226,22 @@ def tica(data=None, lag=10, dim=-1, var_cutoff=0.95, kinetic_map=True, commute_m
                weights=weights, reversible=reversible, ncov_max=ncov_max)
-    return _param_stage(data, res, stride=stride)
+    if data is not None:
+        res.estimate(data, chunksize=chunk_size)
```
`chunk_size`

clonker commented on pyemma/coordinates/api.py:
```diff
@@ -1239,20 +1226,22 @@ def tica(data=None, lag=10, dim=-1, var_cutoff=0.95, kinetic_map=True, commute_m
 
 def covariance_lagged(data=None, c00=True, c0t=True, ctt=False, remove_constant_mean=None, remove_data_mean=False,
-                      reversible=False, bessel=True, lag=0, weights="empirical", stride=1, skip=0, chunksize=1000):
+                      reversible=False, bessel=True, lag=0, weights="empirical", stride=1, skip=0, chunksize=None):
```
`chunk_size`

clonker commented on pyemma/coordinates/api.py:
```diff
@@ -1310,7 +1303,7 @@ def covariance_lagged(data=None, c00=True, c0t=True, ctt=False, remove_constant_
                 raise ValueError("Data must be supplied for reweighting='koopman'")
             koop = _KoopmanEstimator(lag=lag, stride=stride, skip=skip)
-            _param_stage(data, koop, stride=stride)
+            koop.estimate(data, chunksize=chunksize)
```
`chunk_size`

clonker commented on pyemma/coordinates/api.py:
```diff
@@ -1323,10 +1316,13 @@ def covariance_lagged(data=None, c00=True, c0t=True, ctt=False, remove_constant_
-    return _param_stage(data, lc, stride=stride)
+                          weights=weights, stride=stride, skip=skip)
+    if data is not None:
+        lc.estimate(data, chunksize=chunksize)
```
`chunk_size`

clonker commented on pyemma/coordinates/api.py:
```diff
@@ -1376,12 +1372,14 @@ def cluster_mini_batch_kmeans(data=None, k=100, max_iter=10, batch_size=0.2, met
                                     batch_size=batch_size, n_jobs=n_jobs, skip=skip, clustercenters=clustercenters)
-    return _param_stage(data, res, stride=1, chunk_size=chunk_size)
+    if data is not None:
+        res.estimate(data, chunksize=chunk_size)
```
`chunk_size`

clonker commented on pyemma/coordinates/api.py:
```diff
@@ -1504,12 +1504,14 @@ def cluster_kmeans(data=None, k=None, max_iter=10, tolerance=1e-5, stride=1,
-    return _param_stage(data, res, stride=stride, chunk_size=chunk_size)
+                           keep_data=keep_data, clustercenters=clustercenters, stride=stride)
+    if data is not None:
+        res.estimate(data, chunksize=chunk_size)
```
`chunk_size`

clonker commented on pyemma/coordinates/api.py:
```diff
@@ -1575,8 +1579,10 @@ def cluster_uniform_time(data=None, k=None, stride=1, metric='euclidean',
-    return _param_stage(data, res, stride=stride, chunk_size=chunk_size)
+    res = UniformTimeClustering(k, metric=metric, n_jobs=n_jobs, skip=skip, stride=stride)
+    if data is not None:
+        res.estimate(data, chunksize=chunk_size)
```
`chunk_size`

clonker commented on pyemma/coordinates/api.py:
```diff
@@ -1669,7 +1677,9 @@ def cluster_regspace(data=None, dmin=-1, max_centers=1000, stride=1, metric='euc
                                   n_jobs=n_jobs, stride=stride, skip=skip)
-    return _param_stage(data, res, stride=stride, chunk_size=chunk_size)
+    if data is not None:
+        res.estimate(data, chunksize=chunk_size)
```
`chunk_size`

clonker commented on pyemma/coordinates/api.py:
```diff
@@ -1757,9 +1769,9 @@ def assign_to_centers(data=None, centers=None, stride=1, return_dtrajs=True,
     from pyemma.coordinates.clustering.assign import AssignCenters
     res = AssignCenters(centers, metric=metric, n_jobs=n_jobs, skip=skip)
+    if data is not None:
+        res.estimate(data, chunksize=chunk_size)
```
`chunk_size`

clonker commented on pyemma/coordinates/data/_base/datasource.py:
```diff
@@ -34,7 +34,7 @@ class DataSource(Iterable, TrajectoryRandomAccessible):
     """
 
-    def __init__(self, chunksize=1000):
+    def __init__(self, chunksize=None):
```
`chunk_size`

clonker commented on pyemma/coordinates/data/_base/iterable.py:
```diff
@@ -57,7 +59,9 @@ def chunksize(self):
     @chunksize.setter
     def chunksize(self, value):
-        self._default_chunksize = value
+        self._default_chunksize = value if value is not None else DEFAULT_CHUNKSIZE
```
perhaps rename this to `chunk_size`

clonker commented on pyemma/coordinates/data/_base/streaming_estimator.py:
```diff
@@ -32,36 +31,19 @@ class StreamingEstimator(Estimator):
-        self._chunksize = chunksize
-
-    def estimate(self, X, **kwargs):
+    def estimate(self, X, chunksize=None, **kwargs):
```
`chunk_size`

clonker commented on pyemma/coordinates/data/_base/transformer.py:
```diff
@@ -191,10 +191,9 @@ def chunksize(self):
-        self.data_producer.chunksize = int(size)
+        if self.data_producer is None:
+            raise RuntimeError('cant set chunksize')
+        self.data_producer.chunksize = size
```
`chunk_size`

clonker commented on pyemma/coordinates/data/data_in_memory.py:
```diff
@@ -50,7 +50,7 @@ class DataInMemory(DataSource):
         return DataInMemoryIterator(self, skip, chunk, stride, return_trajindex, cols)
 
-    def __init__(self, data, chunksize=5000, **kw):
+    def __init__(self, data, chunksize=None, **kw):
```
`chunk_size`

clonker commented on pyemma/coordinates/estimation/covariance.py:
```diff
@@ -77,14 +77,14 @@ class LaggedCovariance(StreamingEstimator):
          skip the first initial n frames per trajectory.
-     chunksize : int, optional, default=None
-         The chunk size at which the input files are being processed.
+     chunksize : deprecated, default=NoTImplemented
```
`chunk_size`

clonker commented on pyemma/coordinates/estimation/covariance.py:
```diff
@@ -77,14 +77,14 @@ class LaggedCovariance(StreamingEstimator):
                  bessel=True, sparse_mode='auto', modify_data=False, lag=0, weights=None, stride=1, skip=0,
-                 chunksize=None, ncov_max=float('inf')):
-        super(LaggedCovariance, self).__init__(chunksize=chunksize)
+                 chunksize=NotImplemented, ncov_max=float('inf')):
```
`chunk_size`

clonker commented on pyemma/coordinates/estimation/covariance.py:
```diff
@@ -166,14 +166,15 @@ def _estimate(self, iterable, partial_fit=False):
         self.logger.debug("will use %s total frames for %s",
                           iterable.trajectory_lengths(self.stride, skip=self.skip), self.name)
 
+        chunksize = 0 if partial_fit else iterable.chunksize
```
`chunk_size`

clonker commented on pyemma/coordinates/estimation/covariance.py:
```diff
@@ -166,14 +166,15 @@ def _estimate(self, iterable, partial_fit=False):
+        chunksize = 0 if partial_fit else iterable.chunksize
         it = iterable.iterator(lag=self.lag, return_trajindex=False, stride=self.stride, skip=self.skip,
-                               chunk=self.chunksize if not partial_fit else 0)
+                               chunk=chunksize)
```
`chunk_size`

clonker commented on pyemma/coordinates/estimation/covariance.py:
```diff
@@ -166,14 +166,15 @@ def _estimate(self, iterable, partial_fit=False):
                 self.weights.data_producer = iterable
             it_weights = self.weights.iterator(lag=0, return_trajindex=False, stride=self.stride, skip=self.skip,
-                                               chunk=self.chunksize if not partial_fit else 0)
+                                               chunk=chunksize)
```
`chunk_size`

clonker commented on pyemma/coordinates/tests/test_eq_covar_estimator.py:
```diff
@@ -34,8 +34,8 @@ def setUpClass(cls):
-        Kest = _KoopmanEstimator(cls.tau, epsilon=cls.epsilon, chunksize=cls.chunksize)
-        Kest.estimate(cls.source_obj)
+        Kest = _KoopmanEstimator(cls.tau, epsilon=cls.epsilon)
+        Kest.estimate(cls.source_obj, chunksize=cls.chunksize)
```
`chunk_size`

clonker commented on pyemma/coordinates/tests/test_koopman_estimator.py:
```diff
@@ -87,8 +87,8 @@ def setUpClass(cls):
-        cls.K_est = _KoopmanEstimator(cls.tau, epsilon=cls.epsilon, chunksize=cls.chunksize)
-        cls.K_est.estimate(cls.source_obj)
+        cls.K_est = _KoopmanEstimator(cls.tau, epsilon=cls.epsilon)
+        cls.K_est.estimate(cls.source_obj, chunksize=cls.chunksize)
```
`chunk_size`

clonker commented on pyemma/coordinates/tests/test_pipeline.py:
```diff
@@ -193,7 +193,7 @@ def test_chunksize(self):
         chunksize = 1001
         chain = [reader_xtc, api.tica(), api.cluster_mini_batch_kmeans(batch_size=0.3, k=3)]
-        p = api.pipeline(chain, chunksize=chunksize)
+        p = api.pipeline(chain, chunksize=chunksize, run=False)
```
`chunk_size`

clonker reviewed (commented):
I like the changes, the code in itself becomes a lot cleaner (especially in the API).  :)

There is some inconsistency between `chunksize` and `chunk_size` and I would probably opt for sticking with one of them and eliminating the other.
The default chunk size should probably come from the config file, as we have already discussed.

marscher commented:
thanks for the review @clonker. The chunksize vs. chunk_size thing can be solved without breaking the api or deprecating first. I think chunk_size is only used in the API, but was never used in any estimator. Actually chunk_size is more appropriate, but deprecating and switching over would break a lot of code IMO. Agreed about the misleading usage in Iterable.

Putting the default value in the config would be nice as well. We also discussed on accepting strings to directly limit the amount of memory a chunk can occupy. Do you suggest to use pint to do the conversion, which has the nice property we could accept SI units for memory, or shall we hack something custom for this?

clonker commented:
True, I didn't consider the breaking api thing. 
Wrt units: I think pint is the way to go, especially if one would need other units / conversions in the future. However it does mean to introduce a new dependency. But then again that dependency is pure python, so it shouldn't cause much trouble.

marscher committed:
[api] pass stride in assign_to_centers

marscher committed:
[coordinates] made default chunksize a configuration variable

This is a human readable string (eg. '10m') variable in pyemma.config

All coordinate streamed classes by default use None as chunksize and
determine the value from the config then.

current default = 2m

It is implemented by obtaining the output type and the dimension.

marscher commented:
didnt go for pint but added the reverse function to bytes_to_string ;)

marscher committed:
[doc] amend changelog [ci skip]

marscher committed:
fix test

marscher commented:
thanks!
